### PR TITLE
Bump Boost minimum version to 1.71.0

### PR DESCRIPTION
**Pull request by aurianer**
This allows to replace spirit.qi with x3
https://github.com/STEllAR-GROUP/hpx/pull/5092#issuecomment-751365759

Any objections?

### CODE REVIEW

**Comment by msimberg** [comment]
> This allows to replace spirit.qi with x3
> [#5092 (comment)](https://github.com/STEllAR-GROUP/hpx/pull/5092#issuecomment-751365759)
> 
> Any objections?

No objections at all, but CI will need some updates...

File: libs/full/program_options/docs/index.rst [hunk 1]
```diff
@@ -12,8 +12,8 @@ program_options
 ===============
 
 The module program_options is a direct fork of the Boost.ProgramOptions library
-(Boost V1.70.0). For more information about this library please see `here
-<https://www.boost.org/doc/libs/1_70_0/doc/html/program_options.html>`__.
+(Boost V1.71.0). For more information about this library please see `here
+<https://www.boost.org/doc/libs/1_71_0/doc/html/program_options.html>`__.
```
Comment (msimberg) [inline libs/full/program_options/docs/index.rst:16]: Hmm, it is still based on 1.70 even if we require 1.71. I'd either keep the note about 1.70, or remove it altogether.

Comment (hkaiser) [reply libs/full/program_options/docs/index.rst:16]: I'd like to see this note left in place (while referring to the original V1.70.0).

**Comment by aurianer** [comment]
Waiting to have the corresponding Boost modules/files on Daint

**Comment by aurianer** [comment]
This PR will fail until https://github.com/eth-cscs/production/pull/2179 is merged

**Comment by aurianer** [comment]
I think I've addressed all the CI related issues, let's see what's the next run says..
`env-icc` jenkins run is failing but this problem will be addressed in #5199.

**Comment by msimberg** [comment]
Waiting for a reply to https://github.com/boostorg/math/issues/181#issuecomment-792766476 before doing anything with the Cray clang builder, but we may have to disable/change it for the moment.

**Comment by msimberg** [comment]
I've added a custom Boost config which unconditionally disables `BOOST_HAVE_FLOAT128` on the Cray clang builder for now. There may be a better workaround/fix in the future. Assuming I managed to get the syntax right for the CMake scripts this is ready to go. I'm ignoring the ICC failures as they'll be fixed separately.

**Pull request merged**